### DISCUSSION TITLE
Building with meson/ninja seems to cause glib clash with other libs while building with make doesn't

### DISCUSSION THREAD

**Discussion by tri-bao** [libvips/libvips#3845] [General]
I have a simple .NETCore program using just 2 libs
```
        <PackageReference Include="NetVips" Version="2.3.0" />
        <PackageReference Include="Magick.NET-Q16-AnyCPU" Version="12.1.0" />
```

`Magick.NET` includes everything it needs in its own static library. `NetVips` load `libvips` from the runtime system which is on linux x64, ubuntu.

The program uses `NetVips` (hence `libvips`) to convert files form different formats to tiff. Then it uses `Magick.NET` to combine all tiff files into a single tiff file.

Everything works fine with libvips `8.12.x` series except that we can't use `tiffsave_target` action which is only available from libvips `8.13.x` series. So, I had to build libvips from source as the ubuntu 22.04 package manager only delivers libvips 8.12

libvips `8.13` supports both `meson` and `make`.

When I used `meson`, my program crashed with the following error when `Magick.NET` is called (reminding that `NetVips` is called befere `NetVips`)

```
(process:1): GLib-GObject-WARNING **: 03:35:20.257: cannot register existing type 'gchar'
**
GLib-GObject:ERROR:../gobject/gvaluetypes.c:455:_g_value_types_init: assertion failed: (type == G_TYPE_CHAR)
Bail out! GLib-GObject:ERROR:../gobject/gvaluetypes.c:455:_g_value_types_init: assertion failed: (type == G_TYPE_CHAR)
```
When I used `make`, my program worked without any error.

Starting from version `8.14.x`, libvips only support `meson`. I tried to build all `8.14.x` series and `8.15.x` series. Both caused my program crashed with the same error above. Thus now, we need to stick with version `8.13.3` and couldn't upgrade to later versions.

Could you help check if it's due to I made something terribly wrong in my build script?

**My build script below**

```
DEBIAN_FRONTEND=noninteractive \
    && BUILD_DEPS="curl \
      build-essential \
      pkg-config \
      bc \
      meson \
      ninja-build \
      libglib2.0-dev \
      \
      libfftw3-dev \
      libopenexr-dev \
      libgsf-1-dev \
      liborc-0.4-dev \
      libopenslide-dev \
      libmatio-dev \
      libwebp-dev \
      libjpeg-turbo8-dev \
      libexpat1-dev \
      libexif-dev \
      libtiff5-dev \
      libcfitsio-dev \
      libpoppler-glib-dev \
      librsvg2-dev  \
      libpango1.0-dev \
      libopenjp2-7-dev \
      libimagequant-dev \
      \
      libpng-dev \
      libxml2-dev \
      libcairo2-dev \
      libgirepository1.0-dev \
      libheif-dev \
      libgif-dev \
      libarchive-dev \
      libcgif-dev \
      libfontconfig1-dev \
      \
      libnifti-dev \
      libhwy-dev \
      liblcms2-dev \
      libffi-dev \
    " \
    && apt-get install -y --no-install-recommends \
      $BUILD_DEPS \
          libgdiplus\
          libc6-dev \
          libc6 \
          libicu-dev \
          libfontconfig1 \
          fontconfig

# version 8.13 Still supports make and delivered file tar.gz
# Starting from 8.14, only meson is supported and delivered file tar.xz
LIBVIPS_VERSION='8.13.3'
LIBVIPS_TAR='tar.gz'
LIBVIPS_UNTAR='xzf'

curl -Lo /tmp/vips-${LIBVIPS_VERSION}.${LIBVIPS_TAR} https://github.com/libvips/libvips/releases/download/v${LIBVIPS_VERSION}/vips-${LIBVIPS_VERSION}.${LIBVIPS_TAR}
```

Below build steps are using `meson` and will cause the crash
```
cd /tmp \
    && tar -${LIBVIPS_UNTAR} vips-${LIBVIPS_VERSION}.${LIBVIPS_TAR} \
    && cd vips-${LIBVIPS_VERSION} \
    && meson setup build -Dmagick=disabled --buildtype=release \
    && cd build \
    && ninja \
    && ninja install \
    && ldconfig \
    && vips --vips-version \
    && vips --vips-config
```

Below build steps are using `make` and will work without any error
```
cd /tmp \
    && tar -${LIBVIPS_UNTAR} vips-${LIBVIPS_VERSION}.${LIBVIPS_TAR} \
    && cd vips-${LIBVIPS_VERSION} \
    && ./configure --without-magick \
    && make -j4 \
    && make install \
    && ldconfig \
    && vips --vips-version \
    && vips --vips-config
```

**Comment by lovell** [comment]
> (process:1): GLib-GObject-WARNING **: 03:35:20.257: cannot register existing type 'gchar'

This message can appear when there are multiple conflicting versions of the glib shared library on a system, which hopefully provides you with a bit more information about where you can investigate next.

**Comment by kleisauke** [comment]
This looks like a issue in the prebuilt binaries provided by Magick.NET.
`TestProgram.csproj`:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <Platform>x64</Platform>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Magick.NET-Q16-AnyCPU" Version="13.5.0" />
  </ItemGroup>
</Project>
```

`Program.cs`:
```csharp
using ImageMagick;

using var image = new MagickImage();
```

```console
$ LD_PRELOAD=/lib64/libgobject-2.0.so.0 dotnet run -c Release

(process:17428): GLib-GObject-CRITICAL **: 11:21:49.399: cannot register existing type 'gchar'
**
GLib-GObject:ERROR:../gobject/gvaluetypes.c:455:_g_value_types_init: assertion failed: (type == G_TYPE_CHAR)
Bail out! GLib-GObject:ERROR:../gobject/gvaluetypes.c:455:_g_value_types_init: assertion failed: (type == G_TYPE_CHAR)
```

They probably need to link their binaries with `-Wl,-Bsymbolic-functions` to avoid such problems on Linux, see:
https://github.com/lovell/sharp-libvips/issues/60

**Comment by tri-bao** [comment]
Thanks @lovell , @kleisauke  for explanations. I'll contact MagickNet.

There is still something that isn't clear for me: the same source code 8.13.3, why building with `make` doesn't  lead to the `GLib-GObject:ERROR` while building with `meson` does? Is that the `make` build use `-Wl,-Bsymbolic-functions` but `meson` build doesn't use it?
